File: internal/application/queries/list_lore_by_scope.go
package queries

import (
	"context"

	appcontext "github.com/memlore/memlore/internal/application/context"
	"github.com/memlore/memlore/internal/application/ports"
	"github.com/memlore/memlore/internal/domain"
)

// ListLoreByScopeQuery lists lore for an exact scope.
type ListLoreByScopeQuery struct {
	Scope        domain.Scope
	IncludeStale bool
}

// ListLoreByScopeHandler lists lore entries for an exact scope.
type ListLoreByScopeHandler struct {
	begin ports.UnitOfWorkFactory
}

func NewListLoreByScopeHandler(begin ports.UnitOfWorkFactory) *ListLoreByScopeHandler {
	return &ListLoreByScopeHandler{begin: begin}
}

func (h *ListLoreByScopeHandler) Handle(ctx context.Context, query ListLoreByScopeQuery) ([]domain.LoreEntry, error) {
	uow, err := h.begin(ctx)
	if err != nil {
		return nil, err
	}
	defer uow.Rollback(ctx)
	items, err := uow.LoreEntries().ListByScope(ctx, query.Scope)
	if err != nil {
		return nil, err
	}
	if query.IncludeStale {
		return items, nil
	}
	return appcontext.FilterCurrent(items), nil
}

File: src/memlore/application/queries/list_lore_by_scope.py
from __future__ import annotations

from memlore.application.ports.unit_of_work import UnitOfWork
from memlore.domain.models.lore_entry import LoreEntry
from memlore.domain.models.scope import Scope


class ListLoreByScopeHandler:
    def __init__(self, uow: UnitOfWork) -> None:
        self._uow = uow

    def handle(self, scope: Scope) -> list[LoreEntry]:
        return self._uow.lore_entries.list_by_scope(scope)
